apps/dcos_overlay/src/dcos_overlay_poller.erl

```diff
@@ -171,19 +173,75 @@ maybe_create_vtep(_Overlay = #mesos_state_agentoverlayinfo{backend = Backend},
     VTEPNameStr = binary_to_list(VTEPName),
     ParsedVTEPMAC = list_to_tuple(parse_vtep_mac(VTEPMAC)),
     {ParsedVTEPIP, PrefixLen} = parse_subnet(VTEPIP),
-    dcos_overlay_netlink:iplink_add(Pid, VTEPNameStr, "vxlan", VNI, 64000),
-    {ok, _} = dcos_overlay_netlink:iplink_set(Pid, ParsedVTEPMAC, VTEPNameStr),
-    {ok, _} = dcos_overlay_netlink:ipaddr_replace(Pid, inet, ParsedVTEPIP, PrefixLen, VTEPNameStr),
-    case {application:get_env(dcos_overlay, enable_ipv6, true), VTEPIP6} of
+    case vtep_present(Pid, VXLan) of
+        false ->
+            lager:debug("~p will be created.", [VTEPNameStr]),
+            dcos_overlay_netlink:iplink_add(Pid, VTEPNameStr, "vxlan", VNI, 64000),
+            {ok, _} = dcos_overlay_netlink:iplink_set(Pid, ParsedVTEPMAC, VTEPNameStr);
+        {true, match} ->
+            lager:debug("Attributes of ~p is up-to-date.", [VTEPNameStr]);
+        {true, notmatch} ->
+            lager:info("Attributes of ~p are not update-to-date, and vtep will be recreated.", [VTEPNameStr]),
+            dcos_overlay_netlink:iplink_delete(Pid, VTEPNameStr),
+            dcos_overlay_netlink:iplink_add(Pid, VTEPNameStr, "vxlan", VNI, 64000),
+            {ok, _} = dcos_overlay_netlink:iplink_set(Pid, ParsedVTEPMAC, VTEPNameStr)
+    end,
+
+    config_l3(VTEPNameStr, ParsedVTEPIP, PrefixLen, VTEPIP6, State).
+
+config_l3(IfName, IPAddressV4, PrefixLen, IPAddressV6, #state{netlink = Pid}) ->
+    {ok, _} = dcos_overlay_netlink:ipaddr_replace(Pid, inet, IPAddressV4, PrefixLen, IfName),
+    case {application:get_env(dcos_overlay, enable_ipv6, true), IPAddressV6} of
       {false, _} ->
           ok;
       {_, undefined} ->
           ok;
       _ ->
-          ok = try_enable_ipv6(VTEPNameStr),
-          {ParsedVTEPIP6, PrefixLen6} = parse_subnet(VTEPIP6),
-          {ok, _} = dcos_overlay_netlink:ipaddr_replace(Pid, inet6, ParsedVTEPIP6, PrefixLen6, VTEPNameStr)
+          ok = try_enable_ipv6(IfName),
+          {ParsedVTEPIP6, PrefixLen6} = parse_subnet(IPAddressV6),
+          {ok, _} = dcos_overlay_netlink:ipaddr_replace(Pid, inet6, ParsedVTEPIP6, PrefixLen6, IfName)
+    end.
+
+vtep_present(Pid, VXLan) ->
+    #mesos_state_vxlaninfo{vtep_name = VTEPName} = VXLan,
+    VTEPNameStr = binary_to_list(VTEPName),
+    case dcos_overlay_netlink:iplink_show(Pid, VTEPNameStr) of
+        {ok, ?LINK_INFO(LinkInfo)} ->
+            Match = vxlan_attributes_match(LinkInfo, VXLan),
+            {true, Match};
+        {error, ErrorCode, ResponseMsg} ->
+            lager:error("Failed to find ~p for error_code: ~p, msg: ~p", [VTEPNameStr, ErrorCode, ResponseMsg]),
+            false;
+        {ok, []} ->
```

Comment:
Could you please explain me this case? Why does it return `{true, match}` here?

Reply:
In fact, {ok, []} will never occur in normal circustance except that we are running dcos-net with non-root user, as in ource circle test with `circleci` user.
```
netlink_request(Pid, Type, Flags, Msg) ->
  Uid = list_to_integer(string:strip(os:cmd("id -u"), right, $\n)),
  case Uid of
    0 -> %% root
      gen_netlink_client:rtnl_request(Pid, Type, Flags, Msg);
    _ ->
      %% here here
      io:format("Would run fun ~p with flags ~p and argument ~p~n", [Type, Flags, Msg]),
      {ok, []}
  end.
```
I can add comment on that for explanation, but it's better to support non-root user to run dcos-net for security.

Reply:
It's not "for security". I'd prefer to remove [1] and disable all overlay tests if non-root user starts them. That's what we are doing in l4lb common tests [2].

@GoelDeepak what would you say?

[1] https://github.com/dcos/dcos-net/blob/08ca22f/apps/dcos_overlay/src/dcos_overlay_netlink.erl#L176-L194
[2] https://github.com/dcos/dcos-net/blob/ce621b7/apps/dcos_l4lb/test/dcos_l4lb_netlink_SUITE.erl#L41